File: 算法设计与分析/实验/算法5/lab5_bridge.py
from collections import deque
from time import perf_counter
import sys


def read_graph(filename):
    with open(filename, "r", encoding="utf-8") as f:
        n = int(f.readline())
        m = int(f.readline())
        edges = []
        for _ in range(m):
            u, v = map(int, f.readline().split())
            edges.append((u, v))
    return n, edges


def build_adj(n, edges):
    adj = [[] for _ in range(n)]
    for i, (u, v) in enumerate(edges):
        adj[u].append((v, i))
        adj[v].append((u, i))
    return adj


def reachable_without_edge(n, adj, source, target, removed_edge):
    visited = bytearray(n)
    stack = [source]
    visited[source] = 1
    while stack:
        u = stack.pop()
        if u == target:
            return True
        for v, edge_id in adj[u]:
            if edge_id == removed_edge or visited[v]:
                continue
            visited[v] = 1
            stack.append(v)
    return False


def bridges_baseline(n, edges):
    adj = build_adj(n, edges)
    bridges = []
    for i, (u, v) in enumerate(edges):
        if not reachable_without_edge(n, adj, u, v, i):
            bridges.append(i)
    return bridges


class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a, b):
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.rank[ra] < self.rank[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        if self.rank[ra] == self.rank[rb]:
            self.rank[ra] += 1
        return True


class TreePathSkipper:
    def __init__(self, parent):
        self.parent = parent
        self.jump = list(range(len(parent)))

    def find(self, x):
        root = x
        while self.jump[root] != root:
            root = self.jump[root]
        while self.jump[x] != x:
            nxt = self.jump[x]
            self.jump[x] = root
            x = nxt
        return root

    def skip_to_parent(self, x):
        if self.parent[x] != -1:
            self.jump[x] = self.find(self.parent[x])


def bridges_fast(n, edges):
    dsu = DSU(n)
    tree = [[] for _ in range(n)]
    is_tree_edge = bytearray(len(edges))
    is_non_bridge = bytearray(len(edges))
    non_tree_edges = []

    for i, (u, v) in enumerate(edges):
        if dsu.union(u, v):
            is_tree_edge[i] = 1
            tree[u].append((v, i))
            tree[v].append((u, i))
        else:
            non_tree_edges.append(i)
            is_non_bridge[i] = 1

    parent = [-1] * n
    depth = [0] * n
    parent_edge = [-1] * n
    visited = bytearray(n)

    for root in range(n):
        if visited[root]:
            continue
        visited[root] = 1
        q = deque([root])
        while q:
            u = q.popleft()
            for v, edge_id in tree[u]:
                if visited[v]:
                    continue
                visited[v] = 1
                parent[v] = u
                depth[v] = depth[u] + 1
                parent_edge[v] = edge_id
                q.append(v)

    skipper = TreePathSkipper(parent)

    for edge_id in non_tree_edges:
        u, v = edges[edge_id]
        a, b = skipper.find(u), skipper.find(v)
        while a != b:
            if depth[a] < depth[b]:
                a, b = b, a
            pe = parent_edge[a]
            if pe == -1:
                break
            is_non_bridge[pe] = 1
            skipper.skip_to_parent(a)
            a, b = skipper.find(a), skipper.find(b)

    return [i for i in range(len(edges)) if is_tree_edge[i] and not is_non_bridge[i]]


def sample_graph2():
    return 16, [
        (0, 1), (0, 2), (1, 2),
        (2, 3),
        (3, 4), (3, 5), (4, 5),
        (5, 6),
        (6, 7), (7, 8), (8, 9), (9, 6),
        (9, 10),
        (10, 11), (10, 12), (11, 12),
        (12, 13),
        (13, 14),
        (14, 15),
    ]


def run(name, n, edges, algorithm):
    start = perf_counter()
    bridges = algorithm(n, edges)
    seconds = perf_counter() - start
    shown = " ".join(f"({edges[i][0]},{edges[i][1]})" for i in bridges[:20])
    if len(bridges) > 20:
        shown += " ..."
    print(name)
    print(f"vertices: {n}, edges: {len(edges)}")
    print(f"bridge count: {len(bridges)}")
    print(f"time: {seconds:.6f} s")
    print(f"bridges: {shown}\n")
    return bridges, seconds


def main():
    if len(sys.argv) == 1:
        n, edges = sample_graph2()
        base, _ = run("sample graph2 - baseline", n, edges, bridges_baseline)
        fast, _ = run("sample graph2 - fast dsu", n, edges, bridges_fast)
        print("same result:", "yes" if base == fast else "no")
        return

    filename = sys.argv[1]
    mode = sys.argv[2] if len(sys.argv) >= 3 else "fast"
    n, edges = read_graph(filename)
    if mode == "baseline":
        run(filename + " - baseline", n, edges, bridges_baseline)
    elif mode == "both":
        base, _ = run(filename + " - baseline", n, edges, bridges_baseline)
        fast, _ = run(filename + " - fast dsu", n, edges, bridges_fast)
        print("same result:", "yes" if base == fast else "no")
    else:
        run(filename + " - fast dsu", n, edges, bridges_fast)


if __name__ == "__main__":
    main()

File: 算法设计与分析/实验/算法5/lab5_bridge.cpp
#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <numeric>
#include <queue>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

struct Edge {
    int u;
    int v;
};

class DSU {
public:
    explicit DSU(int n = 0) { reset(n); }

    void reset(int n) {
        parent.resize(n);
        rank.assign(n, 0);
        iota(parent.begin(), parent.end(), 0);
    }

    int find(int x) {
        if (parent[x] == x) return x;
        parent[x] = find(parent[x]);
        return parent[x];
    }

    bool unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (rank[ra] < rank[rb]) swap(ra, rb);
        parent[rb] = ra;
        if (rank[ra] == rank[rb]) rank[ra]++;
        return true;
    }

private:
    vector<int> parent;
    vector<int> rank;
};

struct Graph {
    int n = 0;
    vector<Edge> edges;
};

Graph readGraph(const string& filename) {
    ifstream in(filename);
    if (!in) {
        throw runtime_error("cannot open file: " + filename);
    }

    Graph g;
    int m;
    in >> g.n >> m;
    g.edges.reserve(m);
    for (int i = 0; i < m; ++i) {
        int u, v;
        in >> u >> v;
        g.edges.push_back({u, v});
    }
    return g;
}

vector<vector<pair<int, int>>> buildAdjacency(const Graph& g) {
    vector<vector<pair<int, int>>> adj(g.n);
    for (int i = 0; i < (int)g.edges.size(); ++i) {
        auto [u, v] = g.edges[i];
        adj[u].push_back({v, i});
        adj[v].push_back({u, i});
    }
    return adj;
}

bool reachableWithoutEdge(
    int n,
    const vector<vector<pair<int, int>>>& adj,
    int source,
    int target,
    int removedEdge
) {
    vector<char> visited(n, false);
    vector<int> stack;
    stack.push_back(source);
    visited[source] = true;

    while (!stack.empty()) {
        int u = stack.back();
        stack.pop_back();
        if (u == target) return true;

        for (auto [v, edgeId] : adj[u]) {
            if (edgeId == removedEdge || visited[v]) continue;
            visited[v] = true;
            stack.push_back(v);
        }
    }
    return false;
}

vector<int> findBridgesBaseline(const Graph& g) {
    auto adj = buildAdjacency(g);
    vector<int> bridges;

    for (int i = 0; i < (int)g.edges.size(); ++i) {
        auto [u, v] = g.edges[i];
        if (!reachableWithoutEdge(g.n, adj, u, v, i)) {
            bridges.push_back(i);
        }
    }
    return bridges;
}

class TreePathSkipper {
public:
    explicit TreePathSkipper(const vector<int>& treeParent) : parent(treeParent) {
        jump.resize(parent.size());
        iota(jump.begin(), jump.end(), 0);
    }

    int find(int x) {
        if (jump[x] == x) return x;
        jump[x] = find(jump[x]);
        return jump[x];
    }

    void skipToParent(int x) {
        if (parent[x] != -1) {
            jump[x] = find(parent[x]);
        }
    }

private:
    const vector<int>& parent;
    vector<int> jump;
};

vector<int> findBridgesFast(const Graph& g) {
    DSU component(g.n);
    vector<vector<pair<int, int>>> tree(g.n);
    vector<int> nonTreeEdges;
    vector<char> isTreeEdge(g.edges.size(), false);
    vector<char> isNonBridge(g.edges.size(), false);

    for (int i = 0; i < (int)g.edges.size(); ++i) {
        auto [u, v] = g.edges[i];
        if (component.unite(u, v)) {
            isTreeEdge[i] = true;
            tree[u].push_back({v, i});
            tree[v].push_back({u, i});
        } else {
            nonTreeEdges.push_back(i);
            isNonBridge[i] = true;
        }
    }

    vector<int> parent(g.n, -1), depth(g.n, 0), parentEdge(g.n, -1);
    vector<char> visited(g.n, false);

    for (int root = 0; root < g.n; ++root) {
        if (visited[root]) continue;
        queue<int> q;
        q.push(root);
        visited[root] = true;

        while (!q.empty()) {
            int u = q.front();
            q.pop();
            for (auto [v, edgeId] : tree[u]) {
                if (visited[v]) continue;
                visited[v] = true;
                parent[v] = u;
                depth[v] = depth[u] + 1;
                parentEdge[v] = edgeId;
                q.push(v);
            }
        }
    }

    TreePathSkipper skipper(parent);

    auto markPathAsNonBridge = [&](int u, int v) {
        int a = skipper.find(u);
        int b = skipper.find(v);
        while (a != b) {
            if (depth[a] < depth[b]) swap(a, b);
            int edgeId = parentEdge[a];
            if (edgeId == -1) break;
            isNonBridge[edgeId] = true;
            skipper.skipToParent(a);
            a = skipper.find(a);
            b = skipper.find(b);
        }
    };

    for (int edgeId : nonTreeEdges) {
        auto [u, v] = g.edges[edgeId];
        markPathAsNonBridge(u, v);
    }

    vector<int> bridges;
    for (int i = 0; i < (int)g.edges.size(); ++i) {
        if (isTreeEdge[i] && !isNonBridge[i]) {
            bridges.push_back(i);
        }
    }
    return bridges;
}

Graph sampleGraph2() {
    Graph g;
    g.n = 16;
    g.edges = {
        {0, 1}, {0, 2}, {1, 2},
        {2, 3},
        {3, 4}, {3, 5}, {4, 5},
        {5, 6},
        {6, 7}, {7, 8}, {8, 9}, {9, 6},
        {9, 10},
        {10, 11}, {10, 12}, {11, 12},
        {12, 13},
        {13, 14},
        {14, 15}
    };
    return g;
}

template <typename Func>
pair<vector<int>, double> timed(Func&& func) {
    auto start = chrono::steady_clock::now();
    vector<int> result = func();
    auto end = chrono::steady_clock::now();
    chrono::duration<double> elapsed = end - start;
    return {result, elapsed.count()};
}

void printResult(const string& name, const Graph& g, const vector<int>& bridges, double seconds) {
    cout << name << "\n";
    cout << "vertices: " << g.n << ", edges: " << g.edges.size() << "\n";
    cout << "bridge count: " << bridges.size() << "\n";
    cout << "time: " << seconds << " s\n";
    cout << "bridges:";
    int limit = min<int>(bridges.size(), 20);
    for (int i = 0; i < limit; ++i) {
        const auto& e = g.edges[bridges[i]];
        cout << " (" << e.u << "," << e.v << ")";
    }
    if ((int)bridges.size() > limit) cout << " ...";
    cout << "\n\n";
}

int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    try {
        if (argc < 2) {
            Graph g = sampleGraph2();
            auto [baseBridges, baseTime] = timed([&] { return findBridgesBaseline(g); });
            auto [fastBridges, fastTime] = timed([&] { return findBridgesFast(g); });
            printResult("sample graph2 - baseline", g, baseBridges, baseTime);
            printResult("sample graph2 - fast dsu", g, fastBridges, fastTime);
            return 0;
        }

        string filename = argv[1];
        string mode = argc >= 3 ? argv[2] : "fast";
        Graph g = readGraph(filename);

        if (mode == "baseline") {
            auto [bridges, seconds] = timed([&] { return findBridgesBaseline(g); });
            printResult(filename + " - baseline", g, bridges, seconds);
        } else if (mode == "both") {
            auto [baseBridges, baseTime] = timed([&] { return findBridgesBaseline(g); });
            auto [fastBridges, fastTime] = timed([&] { return findBridgesFast(g); });
            printResult(filename + " - baseline", g, baseBridges, baseTime);
            printResult(filename + " - fast dsu", g, fastBridges, fastTime);
            cout << "same result: " << (baseBridges == fastBridges ? "yes" : "no") << "\n";
        } else {
            auto [bridges, seconds] = timed([&] { return findBridgesFast(g); });
            printResult(filename + " - fast dsu", g, bridges, seconds);
        }
    } catch (const exception& ex) {
        cerr << "error: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
